src/container/rpc.h

```diff
@@ -468,49 +472,558 @@ CRT_RPC_DECLARE(cont_tgt_epoch_aggregate, DAOS_ISEQ_CONT_TGT_EPOCH_AGGREGATE,
 	((d_string_t)		(cadi_principal_name)	CRT_VAR) \
 	((uint8_t)		(cadi_principal_type)	CRT_VAR)
 
+#define DAOS_ISEQ_CONT_ACL_DELETE_V8 /* input fields */                                            \
+	((struct cont_op_v8_in)(cadi_op)CRT_VAR)((d_string_t)(cadi_principal_name)CRT_VAR)(        \
+	    (uint8_t)(cadi_principal_type)CRT_VAR)
+
 #define DAOS_OSEQ_CONT_ACL_DELETE	/* output fields */	 \
 	((struct cont_op_out)	(cado_op)		CRT_VAR)
 
-CRT_RPC_DECLARE(cont_acl_delete, DAOS_ISEQ_CONT_ACL_DELETE,
-		DAOS_OSEQ_CONT_ACL_DELETE)
+CRT_RPC_DECLARE(cont_acl_delete, DAOS_ISEQ_CONT_ACL_DELETE, DAOS_OSEQ_CONT_ACL_DELETE)
+CRT_RPC_DECLARE(cont_acl_delete_v8, DAOS_ISEQ_CONT_ACL_DELETE_V8, DAOS_OSEQ_CONT_ACL_DELETE)
 
 static inline int
-cont_req_create(crt_context_t crt_ctx, crt_endpoint_t *tgt_ep, crt_opcode_t opc,
-		crt_rpc_t **req)
+cont_req_create(crt_context_t crt_ctx, crt_endpoint_t *tgt_ep, crt_opcode_t opc, uuid_t ci_pool_hdl,
+		uuid_t ci_uuid, uuid_t ci_hdl, uint64_t *req_timep, crt_rpc_t **req)
 {
-	crt_opcode_t opcode;
-
-	opcode = DAOS_RPC_OPCODE(opc, DAOS_CONT_MODULE, dc_cont_proto_version ?
-				 dc_cont_proto_version : DAOS_CONT_VERSION);
+	int                    rc;
+	crt_opcode_t           opcode;
+	static __thread uuid_t cli_id;
+	int                    proto_ver;
+	struct cont_op_in     *in;
+
+	if (uuid_is_null(cli_id)) {
+		uuid_generate(cli_id);
+		D_INFO("New client thread: " DF_UUID "\n", DP_UUID(cli_id));
+	}
+
+	proto_ver = dc_cont_proto_version ? dc_cont_proto_version : DAOS_CONT_VERSION;
+	opcode    = DAOS_RPC_OPCODE(opc, DAOS_CONT_MODULE, proto_ver);
 	/* call daos_rpc_tag to get the target tag/context idx */
 	tgt_ep->ep_tag = daos_rpc_tag(DAOS_REQ_CONT, tgt_ep->ep_tag);
 
-	return crt_req_create(crt_ctx, tgt_ep, opcode, req);
+	rc = crt_req_create(crt_ctx, tgt_ep, opcode, req);
+	if (rc != 0)
+		return rc;
+	in = crt_req_get(*req);
+
+	uuid_copy(in->ci_pool_hdl, ci_pool_hdl);
+	uuid_copy(in->ci_uuid, ci_uuid);
+	if (uuid_is_null(ci_hdl))
+		uuid_clear(in->ci_hdl);
+	else
+		uuid_copy(in->ci_hdl, ci_hdl);
+
+	if (req_timep && (*req_timep == 0))
+		*req_timep = d_hlc_get();
+
+	/* Temporary req_timep check: some opcodes aren't (yet) at v8 and don't have the op key */
+	if (req_timep && (proto_ver >= CONT_PROTO_VER_WITH_SVC_OP_KEY)) {
+		struct cont_op_v8_in *in8 = crt_req_get(*req);
+
+		uuid_copy(in8->ci_cli_id, cli_id);
+		in8->ci_time = *req_timep;
+	}
+
+	return rc;
+}
+
+/* utility functions to get/set members with different offsets depending on the protocol version. */
+static inline void
+cont_op_in_get_label(crt_rpc_t *rpc, crt_opcode_t opc, int cont_proto_ver, const char **clbl_out)
+{
+	void *in = crt_req_get(rpc);
+
+	if (cont_proto_ver >= CONT_PROTO_VER_WITH_SVC_OP_KEY) {
+		if (opc == CONT_OPEN_BYLABEL)
+			*clbl_out = ((struct cont_open_bylabel_v8_in *)in)->coli_label;
+		else if (opc == CONT_DESTROY_BYLABEL)
+			*clbl_out = ((struct cont_destroy_bylabel_v8_in *)in)->cdli_label;
+		else
+			*clbl_out = NULL;
```

Comment:
[Question] Would asserting that `opc` cannot be other than the two listed make more sense? Asking because a) we shouldn't call one of these functions if we haven't checked what the `opc` is, and b) we don't seem to be checking if pointers are null after calling these functions. An assertion failure is cleaner than a null-pointer dereference.